deps/rabbit/src/rabbit_fifo.erl

```diff
@@ -604,293 +601,189 @@ apply(Meta, {nodeup, Node}, #?MODULE{consumers = Cons0,
-    MessagesTotal = lqueue:len(MessagesV2) +
-                    lqueue:len(ReturnsV2) +
-                    MessagesConsumersV2 +
-                    MessagesWaitingConsumersV2,
-
-    #?MODULE{cfg = Cfg,
-             messages = MessagesV2,
-             messages_total = MessagesTotal,
-             returns = ReturnsV2,
-             enqueue_count = rabbit_fifo_v1:get_field(enqueue_count, V1State),
-             enqueuers = EnqueuersV2,
-             ra_indexes = IndexesV1,
-             release_cursors = rabbit_fifo_v1:get_field(release_cursors, V1State),
-             consumers = ConsumersV2,
-             service_queue = rabbit_fifo_v1:get_field(service_queue, V1State),
-             msg_bytes_enqueue = rabbit_fifo_v1:get_field(msg_bytes_enqueue, V1State),
-             msg_bytes_checkout = rabbit_fifo_v1:get_field(msg_bytes_checkout, V1State),
-             waiting_consumers = WaitingConsumersV2,
-             last_active = rabbit_fifo_v1:get_field(last_active, V1State)
-            }.
-
-convert_v2_to_v3(#rabbit_fifo{consumers = ConsumersV2} = StateV2) ->
-    ConsumersV3 = maps:map(fun(_, C) ->
-                                   convert_consumer_v2_to_v3(C)
-                           end, ConsumersV2),
-    StateV2#rabbit_fifo{consumers = ConsumersV3}.
-
-convert_consumer_v2_to_v3(C = #consumer{cfg = Cfg = #consumer_cfg{credit_mode = simple_prefetch,
-                                                                  meta = #{prefetch := Prefetch}}}) ->
-    C#consumer{cfg = Cfg#consumer_cfg{credit_mode = {simple_prefetch, Prefetch}}};
-convert_consumer_v2_to_v3(C) ->
-    C.
+convert_v3_to_v4(#{} = _Meta, StateV3) ->
+    %% TODO: consider emitting release cursors as checkpoints
+    Messages0 = rabbit_fifo_v3:get_field(messages, StateV3),
+    Returns0 = lqueue:to_list(rabbit_fifo_v3:get_field(returns, StateV3)),
+    Consumers0 = rabbit_fifo_v3:get_field(consumers, StateV3),
+    Consumers = maps:map(
+                  fun (_, #consumer{checked_out = Ch0} = C) ->
+                          Ch = maps:map(
+                                 fun (_, ?MSG(I, #{delivery_count := DC} = H)) ->
+                                         ?MSG(I, H#{acquired_count => DC});
+                                     (_, Msg) ->
+                                         Msg
+                                 end, Ch0),
+                          C#consumer{checked_out = Ch};
+                      (_, Msg) ->
+                          Msg
+                  end, Consumers0),
+    Returns = lqueue:from_list(
+                lists:map(fun (?MSG(I, #{delivery_count := DC} = H)) ->
+                                  ?MSG(I, H#{acquired_count => DC});
+                              (Msg) ->
+                                  Msg
+                          end, Returns0)),
+
+    Messages = rabbit_fifo_q:from_lqueue(Messages0),
+    #?STATE{cfg = rabbit_fifo_v3:get_field(cfg, StateV3),
+            messages = Messages,
```

Comment:
v3 can set a `delivery_count` in `messages` when `#requeue{}` command is applied.
Looks like we therefore need to convert `delivery_count` to `acquired_count` in `messages` as well?

Reply:
I don't think we need to do this as requeue can be considered a new entry point at which we start counting acquisitions from.